redis/provider_test.go

```diff
@@ -0,0 +1,346 @@
+				ReplicationGroups: []*elasticache.ReplicationGroup{
+					{
+						Status: aws.String("test status"),
+					},
+				},
+			}
+			mockElasticache.DescribeReplicationGroupsWithContextReturns(awsOutput, nil)
+
+			instanceID := "foobar"
+			ctx := context.Background()
+
+			state, stateMessage, stateErr := provider.GetState(ctx, instanceID)
+
+			Expect(state).To(Equal(broker.ServiceState("test status")))
+			Expect(stateMessage).To(Equal("ElastiCache state is test status for cf-qwkec4pxhft6q"))
+			Expect(stateErr).ToNot(HaveOccurred())
+
+			Expect(mockElasticache.DescribeReplicationGroupsWithContextCallCount()).To(Equal(1))
+			passedCtx, passedInput, _ := mockElasticache.DescribeReplicationGroupsWithContextArgsForCall(0)
+			Expect(passedCtx).To(Equal(ctx))
+			Expect(passedInput).To(Equal(&elasticache.DescribeReplicationGroupsInput{
+				ReplicationGroupId: aws.String(replicationGroupID),
+			}))
+		})
+
+		It("handles errors from the AWS API", func() {
+			describeErr := errors.New("some error")
+			mockElasticache.DescribeReplicationGroupsWithContextReturns(nil, describeErr)
+
+			_, _, stateErr := provider.GetState(context.Background(), "foobar")
+			Expect(stateErr).To(MatchError(describeErr))
+		})
+
+		It("returns deleted status for a non-existing replication group", func() {
+			describeErr := awserr.New(elasticache.ErrCodeReplicationGroupNotFoundFault, "some message", nil)
+			mockElasticache.DescribeReplicationGroupsWithContextReturns(nil, describeErr)
+
+			state, stateMessage, stateErr := provider.GetState(context.Background(), "foobar")
+			Expect(state).To(Equal(broker.NonExisting))
+			Expect(stateMessage).To(Equal("Replication group does not exist: cf-qwkec4pxhft6q"))
+			Expect(stateErr).ToNot(HaveOccurred())
+		})
+
+		It("handles an empty replication group list from AWS", func() {
+			describeOutput := &elasticache.DescribeReplicationGroupsOutput{
+				ReplicationGroups: []*elasticache.ReplicationGroup{},
+			}
+			mockElasticache.DescribeReplicationGroupsWithContextReturns(describeOutput, nil)
+			_, _, stateErr := provider.GetState(context.Background(), "foobar")
+			Expect(stateErr).To(MatchError("Invalid response from AWS: no cache clusters returned for cf-qwkec4pxhft6q"))
+		})
+
+		It("handles empty status from AWS", func() {
+			describeOutput := &elasticache.DescribeReplicationGroupsOutput{
+				ReplicationGroups: []*elasticache.ReplicationGroup{{}},
+			}
+			mockElasticache.DescribeReplicationGroupsWithContextReturns(describeOutput, nil)
+			_, _, stateErr := provider.GetState(context.Background(), "foobar")
+			Expect(stateErr).To(MatchError("Invalid response from AWS: status is missing for cf-qwkec4pxhft6q"))
+		})
+
+	})
+
+	Context("when generating replication group names", func() {
+		It("should generate valid and unique values", func() {
+			n := 20000
+			collisionCount := 0
+			instanceIDs := make(map[string]int, n)
+			nameRegexp := regexp.MustCompile("[a-z][a-z0-9\\-]*")
```

Comment:
You could use

~~~go
regexp.MustCompile(`[a-z][a-z0-9\-]*`)
~~~

to prevent escaping.